### ISSUE TITLE
Augment or replace rope with Jedi

### ISSUE THREAD

**Issue by spyder-issues-migrator** [spyder-ide/spyder#1213]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2012-12-30T15:26:39Z_

In the Jedi bookmark of https://code.google.com/r/stevensilvester-rich-text-to-thread/source/browse , I have replaced rope with Jedi.

What Jedi offers is an excellent recursiveness that makes it possible to tease out the proper object from very complex code.  

Jedi appears to be about as fast as rope, unless you are parsing against a large library for the first time that session.  I have engaged with the author of Jedi on how to overcome this.  Also, some objects with dynamic docstrings will not show the same docstring as the rope version, such as functools.partial.

Prior to making the speed improvements, perhaps we can default to rope, and lean on Jedi when rope fails to find the object?

I am using a clone of the main development repo on Ubuntu 12.04, 32bit.

_Original issue: http://code.google.com/p/spyderlib/issues/detail?id=1213_

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2013-01-02T06:10:50Z_

**Status:** Accepted  
  **Labels:** MS-v2.3

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-01-13T14:27:28Z_

Just for an update, I have been using Jedi as a complete replacement for rope for the past few weeks, and have not run into any concerns other than the initial slowness when loading large modules (numpy, wx, PyQt4).  The author of Jedi is almost finished working on a cache method that cuts the initial load time from 3-10 seconds to less than a second.  Also, I pushed all three methods (completion, calltip, and definition) into QThreads, modifying the original AnalysisThread to accommodate the change.  So, the UI does not lock up for any reason.
Finally, I included changes to emulate an ipython terminal experience, including auto-tab, and the navigation keyboard shortcuts.
After all this, I finally feel the editing experience is on par with what I had when using Wing IDE, with the warm fuzziness of open source.

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2013-01-14T16:32:36Z_

These are fantastic news! Once Jedi's cache method is on place, I'd really like to left rope behind! Please let us know when that happens so we can review and merge your work for 2.3.

I don't quite understand what you mean by "ipython terminal experience". Could you elaborate a bit more about it?

**Labels:** Cat-Editor

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-01-14T17:25:01Z_

IPython allows you to hit Tab, and it will provide completions if it makes sense to do so.  I modified the Key_Tab handler to check for a valid completion candidate (python literal, '.', or '(', and no selected text).  It will either provide the completion candidate (or calltip), or indent as usual.  
The other features are the keybindings for navigation, which are the emacs keybindings. These include previous-line, next-line, start-of-line, end-of-line, previous-char, and next-char.  I currently have my configuration set up to use the emacs keybindings for all of these, which required shuffling around some of the defaults.  I also had to catch Ctrl+A, since it is the default "select all" in Qt.  The beginning-of-line moves to the first character on the line, which is not the default emacs behavior, but we could separate them into separate commands if we wanted full emacs compliance.

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2013-01-27T04:52:16Z_

What about jedi maturity?

We really should keep `rope` integrated until we are sure that `jedi` is enough mature/popular. For example, there is no Debian/Ubuntu package for `jedi`.

That being said, I'll be happy to test it.

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-01-27T18:31:18Z_

The author is being very responsive, and I believe we are closing in on a solution to the warm-up problem.  I have encountered very few instances where jedi could not find what I was looking for.  It is available as `pip install jedi`, and has 1700 downloads of the latest version.  My impression is that if we wait for a pypi release incorporating the warm-up changes, we would benefit from a streamlined, simple system by doing a full replacement.  
I did attempt to recreate exactly the call-tip and docstring formats that were being used with rope, and quite honestly with a lot less monkey patching and overall code.  
Please take a test drive and let me know what you think.  Do not forget to switch to the jedi bookmark.
By the way, it is surreal to be having a technical discussion with you Pierre.  Two years ago when I started using Python, I credit you in every way for making that a smooth and enjoyable experience, from Python(x,y) to Spyder to GuiQwt.  I copied the setup.py from spyderlib when creating oct2py, because of the cool way you handled Sphinx.  I have gone from an engineer who writes code when he needs to, to a lifelong open source developer.  Thanks for all you do.

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2013-01-27T18:51:39Z_

@-Steven: I would like to know how Jedi grabs docstrings. Remember that we also use rope to introspect objects on the Editor and send their docstrings to the Object Inspector. So far you haven't mentioned if Jedi can do this and if you have implemented it in your bookmark.

@-Pierre: it seems that jedi has just entered Debian a few days ago. Check this out: http://ftp-master.debian.org/new/python-jedi_0.5~b5-1.html

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-01-27T19:43:42Z_

Yes, jedi is parsing of the raw source code to get the docstring, unless it is a compiled code object.   Currently, functions or classes with decorated or dynamic docstrings do not work properly. I submitted an issue on the jedi Github page.  Otherwise, call tips and the object inspector are working exactly as before.  There are some subtle differences, such as numpy.ones being listed as multiarray.ones in the calltip.  I can create an issue for this if you think it is a problem.

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2013-01-28T01:21:13Z_

@-Carlos: good news for the new python-jedi Debian package. It's a good sign.

@-Steven: thanks for your kind words. I'm always happy to learn that sharing all those projects helped people enjoying Python and its ecosystem!

**Comment by spyder-issues-migrator** [comment]
_From [david.ve...@gmail.com](https://code.google.com/u/108342766018366950755/) on 2013-02-08T01:09:19Z_

It looks interesting to test this although I am not sure how. I haven't worked before with the hg bookmarks. Should I just pull your repository and start using spyder from there?

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-02-08T01:55:46Z_

David, I am going to upload a newer version tonight with some improvements and new features, as well as a better explanation of what those improvements are.  In general, you would clone my repo as usual, then type hg bookmarks.   This gives you a list of bookmarks.   One of them should be called jedi.  Then type hg bookmark jedi.  You can then run bootstrap.py from the spyderlib directory.  You may need to add  --gui=pyqt, I have not had success with Pyside.   

Cheers,
Steve Silvester

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-02-08T18:26:52Z_

Uploaded the changes, and here are the descriptions:

Introspection:
- Switch to jedi due to its enhanced introspection features vs. rope
- Push all jedi operations into QThreads using Threadmanager
  modified Analysis thread to accept variable arguments, renamed to 
  WorkerThread
- Block jedi searches while a search is pending
- Warm up jedi at startup for initial responsiveness
- Use token-based completion/goto if jedi fails to find a match
- Add Enaml file support for completion, goto, and calltip/docstring
- Replace codeeditor.get_primary_at() with non-rope version
Files affected: widgets/sourcecode/codeeditor.py, widgets/editor.py,
     all rope-related files can be removed

Editor:
- Smart Tab key completion (complete or indent)
- Add Enaml file support for syntax highlighting, indent, 
  mark occurences, and code helpers
- Add GoTo definition in context menu in read-only files
- Add emacs-like keyboard shortcuts:
   - beginning of line, end of line
   - prev char, next char
   - prev line, next line
   - select all (to allow the use of Ctrl+A for beginning of line)
Files affected: widgets/sourcecode/codeeditor.py, widgets/editor.py,
    widgets/sourcecode/syntaxhighlighters.py


Way Ahead:
The author of Jedi still feels he needs to make a few changes that will boost speed (startup and completion), and to reach a stable code state.  I ran some speed tests as is, and once jedi has loaded a particular module, it is always faster than rope, and is better at teasing out the right object.  See https://github.com/davidhalter/jedi/issues/102#issuecomment-12981611 for an objective report.

**Comment by spyder-issues-migrator** [comment]
_From [david.ve...@gmail.com](https://code.google.com/u/108342766018366950755/) on 2013-02-09T03:40:28Z_

Steven, thanks for the instructions! Just so to be sure I have it right, jedi is not included in your spyder clone, right? I installed jedi from git (I love Arch Linux, there is a package for virtually everything in the AUR!)

No pyside installed here, always running pyqt

Some early tests from the editor

* tab to invoke the code completion works, but I only get the options that are actually used in the current file, not what's available in the module. Further, following exception is raised:
Traceback (most recent call last):
  File "/home/dave/Software/Development/stevensilvester-rich-text-to-thread/spyderlib/widgets/editor.py", line 237, in update_queue
    end_callback = self.end_callbacks.pop(id(thread))
KeyError: 89345632
note that the key is different for each module I try

* when doing zxc. , and if zxc doesn't exists, zxc. becomes zxc.. after a few seconds
* no baloon tips, no documentation opening after np.zeros(
* the editor is not locked during the second he's looking for stuff (and than raising the exception), seems the threading is working nice from a usability perspective

from the python console
* completion works, documentation is visible, baloon tips are visible
* I don't feel any difference between the normal spyder (rope) and jedi when invoking code completion on pylab, numpy. In both cases it is instantaneous, also on the first run. In de rope spyder editor, a first code completion on pylab takes a few seconds on my system.

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-02-09T06:29:44Z_

David,
Thanks for the feedback, I'm glad at least some of it worked for you.
I had been using Linux Mint for testing, but it does show up in Win XP.  The zxc. problem should be easy to fix.  I have no problem on either platform with numpy.ones.  

What was the version of jedi you installed?  I am using 0.5b.5, as evidenced by calling `pip freeze | grep jedi`, or by looking for a jedi-<version>.egg-info directory in my dist-packages folder.  

Also, you could try running the following snippet from a prompt:
import jedi
code = 'import numpy; numpy.ones('
script = jedi.Script(code, 1, len(code), None)
print script.get_in_function_call()

I get the following object back:
\<CallDef: \<eFunction of \<Function: ones@1791-1830>> index 0>

**Comment by spyder-issues-migrator** [comment]
_From [david.ve...@gmail.com](https://code.google.com/u/108342766018366950755/) on 2013-02-09T08:27:08Z_

the snippet returns:
print script.get_in_function_call()
\<CallDef: \<eFunction of \<Function: ones@120-152>> index 0>

I am not using pip, I installed jedi via my package manager.

My version was build straight from the jedi master git branch, but the last commits are from the beginning of January. Than I switched to the refactor branch since that seems even more fresh, but I get the same behaviour as described in the previous post. The egg file indicates v0.5.5, further:
\>>> jedi.sys.api_version
1013

**Comment by spyder-issues-migrator** [comment]
_From [david.ve...@gmail.com](https://code.google.com/u/108342766018366950755/) on 2013-02-09T08:38:44Z_

Some more trace backs. Although I was using the --debug option, I was not getting a lot of debugging stuff except for the trace back I mentioned. Than I launched spyder as follows:
export SPYDER_DEBUG=True; python2 bootstrap.py -- --debug

And than I see this one:

Traceback (most recent call last):
  File "/home/dave/Software/Development/stevensilvester-rich-text-to-thread/spyderlib/widgets/editor.py", line 175, in run
    self.results = self.function(*self.args)
  File "/home/dave/Software/Development/stevensilvester-rich-text-to-thread/spyderlib/widgets/sourcecode/codeeditor.py", line 79, in warmup
    script.complete()
  File "/usr/lib/python2.7/site-packages/jedi/api.py", line 117, in complete
    names = s.get_defined_names(on_import_stmt=True)
  File "/usr/lib/python2.7/site-packages/jedi/imports.py", line 115, in get_defined_names
    names += self.get_module_names()
  File "/usr/lib/python2.7/site-packages/jedi/imports.py", line 155, in get_module_names
    search_path = self.sys_path_with_modifications()
  File "/usr/lib/python2.7/site-packages/jedi/imports.py", line 167, in sys_path_with_modifications
    parts = self.file_path.split(os.path.sep)
AttributeError: 'NoneType' object has no attribute 'split'

I'll try to figure out what is going on there

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-02-09T08:41:55Z_

David,

Right, I tried the same and it does not work.  Would you mind installing from the pypi tarball? http://pypi.python.org/packages/source/j/jedi/jedi-0.5b5.tar.gz#md5=eaf183537b1fd2eb122b85c526dea729 I'd rather not code against any other versions yet, especially the source, since it is a moving target.

**Comment by spyder-issues-migrator** [comment]
_From [david.ve...@gmail.com](https://code.google.com/u/108342766018366950755/) on 2013-02-09T09:25:10Z_

yes, lets first try to get it to work with one version :-)

And I can confirm now I also have the baloon tips and documentation being displayed while working in the editor (using the 0.5b5 version). So, yes, it works :-) The response is not instant, but the threading makes that the editor is not freezing. If I compare with the rope spyder, I don't feel there is a difference though.

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-02-09T14:47:38Z_

All right, if you pull the latest changes it should address the threading exception, the `zxc.` issue, as well as calltips on builtin modules, which I had temporarily broken in my attempt to get the full calltip path to a module instead of the partial or incorrect path given by jedi (e.g `numeric.ones` -> `numpy.core.ones`).  Thanks again for the feedback.

**Comment by spyder-issues-migrator** [comment]
_From [david.ve...@gmail.com](https://code.google.com/u/108342766018366950755/) on 2013-02-10T03:38:58Z_

Ok, got your latest changes, but I still stay on the 0.5b5 version, right? I quickly checked against the jedi master, dev and refactor branches but there where some errors still there. Let me know when you want more testing against those.

* I lost the baloon tips and documentation. Completion only shows the objects used in the document.

* I assume the debug output intentionally prints the whole current document when jedi loads it?

* one trace back when getting more deeper in to the module structure:
getting results for cv2.cv.
cv2.cv.
[u'cv2.cv.CV_HOUGH_GRADIENT,', u'cv2.cv.BoxPoints(rect)']
Traceback (most recent call last):
  File "/home/dave/Software/Development/stevensilvester-rich-text-to-thread/spyderlib/widgets/editor.py", line 247, in update_queue
    end_callback(thread.results)
  File "/home/dave/Software/Development/stevensilvester-rich-text-to-thread/spyderlib/widgets/editor.py", line 384, in code_completion_finished
    self.trigger_token_completion(automatic)
  File "/home/dave/Software/Development/stevensilvester-rich-text-to-thread/spyderlib/widgets/editor.py", line 375, in trigger_token_completion
    base, obj = obj.split('.')
ValueError: too many values to unpack

(cv2 is OpenCV, an image processing library: http://code.opencv.org/projects/opencv )

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-02-10T10:00:56Z_

Correct, please stick with the pypi version for now.
- I believe I finally licked the threading issues, and made it more robust overall.  
There were two issues there.  
-- The QThread was emitting the `finished` signal but a call to isFinished() returned false.  Now the finished signal supplies its id when finished, so it can be handled explicitly.   
-- Supplying the path argument to jedi.Script() sometimes would not work, but it does provide in-package look-up.  So, I now try with the file path first, then fall back on no path if the introspection fails.
- The whole document should not be printed anymore, I had forgotten to remove a print statement.
- Typing `cv2.cv.` will not result in an error anymore, but jedi does not pick it up.  I will file this as a bug with jedi.  Typing `numpy.lib.` does work, so this is something specific to cv2.

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-02-11T17:53:13Z_

Fixed a few more issues with the latest push.  Jedi was still locking up (my fault, not Jedi's), and token completion with dots still didn't work quite right.  Also, I added token-based goto definition for cython files (works within the file).

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2013-02-11T18:48:47Z_

Impressive work Steven, keep it up!

You know right now we're quite busy finishing 2.2 but we are very interested in your work. As soon as I have some time I'll test it but it seems now David is doing a good job for us! :)

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-02-11T19:07:37Z_

Thanks for the kind works Carlos, v2.2 looks great so far.  David has indeed been a great help.

**Comment by spyder-issues-migrator** [comment]
_From [david.ve...@gmail.com](https://code.google.com/u/108342766018366950755/) on 2013-02-12T13:54:40Z_

I will try to get some more testing done later this week... But thanks Steven for exploring this Jedi route, may the force... Thanks for the thumbs up, but the testing is just a very small effort compared to the coding. So thank you :-)

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-02-14T17:51:47Z_

The latest version is more robust, is refactored and cleaned up, and has better regex based definition finding.  I have been test driving it with several Enaml and Cython files, and there are only a few corner cases where it does not find the right definition. I will continue to work at it (who knew how fun regexes could be?).
Also, I altered `keyPressEvent` for the `CompletionWidget` in `widgets/sourcecode/base.py` to allow Ctrl+P and Ctrl+N to move between completion options (emacs-style keybindings).
Overall, completion and goto are pretty seamless now, even following imports from other Enaml/Cython files.

**Comment by spyder-issues-migrator** [comment]
_From [david.ve...@gmail.com](https://code.google.com/u/108342766018366950755/) on 2013-03-10T04:04:34Z_

Sorry for the delay, but I finally got around doing another round of testing...
* I am still using jedi 0.5b5
* using revision 2410 of your repository
* manually disabled pyqt version check because of issue #1291 (I am on pyqt 4.10)
spyderlib/requirements.py, patch:
@@ -45,10 +45,10 @@
         from spyderlib import qt
         package_name, required_str = qt_infos[qt.API]
         actual_str = qt.__version__
-        if check_version(actual_str, required_str):
-            show_warning("Please check Spyder installation requirements:\n"
-                         "%s %s+ is required (found v%s)."
-                         % (package_name, required_str, actual_str))
+#        if check_version(actual_str, required_str):
+#            show_warning("Please check Spyder installation requirements:\n"
+#                         "%s %s+ is required (found v%s)."
+#                         % (package_name, required_str, actual_str))
     except ImportError:
         show_warning("Please check Spyder installation requirements:\n"
                      "%s %s+ (or %s %s+) is required."

* than I get this traceback (also for the older versions that were still working, so I guess something is wrong on my end, not sure what though):
  File "/home/dave/Software/Development/stevensilvester-rich-text-to-thread/spyderlib/spyder.py", line 1987, in main
    mainwindow = run_spyder(app, options)
  File "/home/dave/Software/Development/stevensilvester-rich-text-to-thread/spyderlib/spyder.py", line 1894, in run_spyder
    main.setup()
  File "/home/dave/Software/Development/stevensilvester-rich-text-to-thread/spyderlib/spyder.py", line 713, in setup
    self.projectexplorer = ProjectExplorer(self)
  File "/home/dave/Software/Development/stevensilvester-rich-text-to-thread/spyderlib/plugins/projectexplorer.py", line 37, in __init__
    self.load_config()
  File "/home/dave/Software/Development/stevensilvester-rich-text-to-thread/spyderlib/plugins/projectexplorer.py", line 128, in load_config
    self.set_workspace(self.get_option('workspace', None))
  File "/home/dave/Software/Development/stevensilvester-rich-text-to-thread/spyderlib/widgets/projectexplorer.py", line 1237, in set_workspace
    self.treewidget.set_workspace(path)
  File "/home/dave/Software/Development/stevensilvester-rich-text-to-thread/spyderlib/widgets/projectexplorer.py", line 645, in set_workspace
    self.workspace.set_root_path(root_path)
  File "/home/dave/Software/Development/stevensilvester-rich-text-to-thread/spyderlib/widgets/projectexplorer.py", line 267, in set_root_path
    self.load()
  File "/home/dave/Software/Development/stevensilvester-rich-text-to-thread/spyderlib/widgets/projectexplorer.py", line 284, in load
    data = cPickle.loads(fdesc.read())
BadPickleGet: 115

**Comment by spyder-issues-migrator** [comment]
_From [david.ve...@gmail.com](https://code.google.com/u/108342766018366950755/) on 2013-03-10T04:08:29Z_

google code automatically puts in the spyder main repo revision, in reality it is of course referring to your repo, link: https://code.google.com/r/stevensilvester-rich-text-to-thread/source/detail?r=35079aacd763dc79dbb5e503184c47a2c15d3281

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-03-10T06:25:32Z_

David, if this is happening across versions, it sounds like your config is corrupted.  Try deleting you config directory (~/.spyder2 on Linux, not sure about OSX),

**Comment by spyder-issues-migrator** [comment]
_From [david.ve...@gmail.com](https://code.google.com/u/108342766018366950755/) on 2013-03-10T07:48:14Z_

My "normal" spyder version (2.2dev) was still working fine, but resetting ~/.spyder2 (I am on Arch Linux) did the trick for use with your repo!

* Smooth experience in both console and editor wrt code completion
* Balloon tips are shown
* Object inspector shows the documentation from either the console or editor (is this actually also done by jedi or another component?)
* In debug mode, all code completion objects are print, but I guess that's intentionally for now
* I do get code completion on scipy.interp. but nothing when using cv2.cv. (cv2, as before, is the OpenCV library). No tracebacks or useful print statements in the console though...

Is there a way to re-trigger the balloon tips? Just like TAB or CTRL+SPACE would trigger the code completion again? It could be handy. Now I would just retype the brackets. If you do CTRL+SPACE or TAB when inside the brackets, the code completion is triggered again. Not sure if possible, but when inside the brackets, it could be handy to show the balloon tips again and/or reload the documentation in the object inspector.

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-03-25T08:33:48Z_

David,
* None of the console code has been changed, that is independent.  
* The cv2.cv is due to it being a compiled library.  I posted an issue against Jedi for that.
* I added the ability to press TAB and re-trigger the calltip, but only if there is a trailing comma.  For example, "functools.partial(func, " + TAB would trigger a balloon tip on functools.partial.  It accounts for opened and closed parens within the definition, such as "functools.partial(func, other_func()," + TAB also works.  
* I will go back an add a keyboard shortcut that will do the same thing, without requiring a trailing comma.

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-03-25T20:59:11Z_

Okay, balloon tips can now be called as follows:
* Press TAB after comma, comma+SPACE, or "("
* Press TAB after an object with no completions (e.g. "functools.partial(oct"+TAB)
* Trigger the "editor > inspect current object" shortcut
All of these will find the previous unclosed parens and give the tip for that object.  (e.g. "functools.partial(func, other_func()," will give the tip for "partial".

Fun little detour, thanks for the suggestion David.

**Comment by spyder-issues-migrator** [comment]
_From [techtonik@gmail.com](https://code.google.com/u/techtonik@gmail.com/) on 2013-06-08T03:42:44Z_

Is it possible to clean up changesets in development line and reupload them to new clone? I'd like to review this feature, but it is hard to do this, and I am not sure that early changesets are still actual.

I saw that rope was just replaced by Jedi. Ideally we should have a choice between systems, paving a road for more implementations.

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-06-09T09:53:01Z_

@-techtonik,  the author of jedi suggested that I wait until v0.6.1 of that library was released to consider it a stable library.  That should be any time now.  Once that happens, I will bring my version up to date with that, then re-baseline on v2.2 of Spyder.  

As to having multiple implementations, how would this be broken out?  Would there be separate plugins for code-completion and goto?  Would there be a prioritized list of plugins that is hard-coded, or would there be a user menu?

**Comment by spyder-issues-migrator** [comment]
_From [techtonik@gmail.com](https://code.google.com/u/techtonik@gmail.com/) on 2013-06-09T15:56:03Z_

It depends on the functionality already provided. I didn't had a chance to look at how deeply rope is hardcoded into Spyder, but there is a lot of stuff in Spyder that is already implemented as plugins, so I don't see problems to do the same for code analysis tools. As to how many plugins will be there - you tell me. =)

**Comment by spyder-issues-migrator** [comment]
_From [jed.lud...@gmail.com](https://code.google.com/u/114333776555678317161/) on 2013-06-17T10:32:39Z_

**Labels:** -MS-v2.3 MS-v2.4

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-10-06T15:11:31Z_

I started from scratch with 2.3.0dev.  I pulled all introspection in a package (utils/introspection).
It will now load a jedi plugin if v0.7.0 or greater is installed, or fall back on rope.

Notes about speed (on my 32bit Windows 7 laptop with 2GB of RAM):
I set jedi up to load numpy at startup, which takes ~10seconds the first time,
and ~1.5seconds at startup thereafter.
Loading PyQt4 takes ~50seconds the first time and ~5seconds on the first trigger per session thereafter.
After that, jedi has been faster and more capable on everything I have thrown at it.

I did sneak in some regex-based fallbacks if either jedi or rope cannot find a definition or completions.
This is also useful when editing non-python-like files. https://code.google.com/p/spyderlib-jedi/source/browse/ Changes:
- added spyderlib\utils\introspection package
   -- uses jedi when available, but falls back to previous rope functions
   -- provides manual fall-back methods when the plugin cannot find something
- spyderlib\plugins\inspector.py -> replaced 'rope' check with introspection check
- spyderlib\widgets\editor.py -> replaced self.rope_project with introspection package calls
- spyderlib\utils\sourcecode.py -> added non-rope-based get_primary_at() and source code manipulation methods
- spyderlib\widgets\sourcecode\codeeditor.py
   -- remove introspection
   -- call sourcecode.get_primary_at()

This does not yet include any of my previous changes that included asynchronous introspection, emacs-like keybindings, super-tab for completion, or support for enaml  files as a python-like file.

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-10-12T13:49:19Z_

All of the base functionality is in the bookmark *jedi-only*.  I added a bookmark *extras* with the following additions:

- Added "Super" Tab to widgets\sourcecode\CodeEditor (Qt.Key_Tab)
- Added enaml support for highlighting and introspection.
  - widgets\sourcecode\codeeditor\CodeEditor
  - widgets\sourcecode\syntaxhighlighter\EnamlSh
- Added emacs keybinding support to widgets\sourcecode\codeeditor\CodeEditor
- Added token-based completion fallback for non-python files and within strings
  - widgets\editor\FileInfo
  - widgets\sourcecode\codeeditor\CodeEditor
  - utils\introspection\get_token_completion_list
- Added enaml icon to spyderlib\images\filetypes based on pyx icon
- Added support for goto definition in non-python-like files
- Added emacs keybindings to spyderlib\widgets\sourcecode\base\CompletionWidget

**Comment by spyder-issues-migrator** [comment]
_From [david.ve...@gmail.com](https://code.google.com/u/108342766018366950755/) on 2013-10-13T13:27:34Z_

Just cloned spyderlib-jedi, but I can't see any bookmarks. I was under the impression bookmarks should come along with clone/update/pull? hg pull -B extras/jedi-only returned a remote bookmark not found. Am I missing something?

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-10-13T16:20:24Z_

Sorry, forgot to explicitly push the bookmarks.

hg clone https://steven.silvester@code.google.com/p/spyderlib-jedi/ 
cd spyderlib-jedi
hg bookmark jedi-only

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-11-03T07:53:55Z_

The latest version allows jedi to provide module completions, including relative imports!  It also includes improvements to the module path finding (without using __import__).  Again, these are all in the jedi-only bookmark.

The extras bookmark has grown substantially and covers many new features:
- Tab completions and re-trigger calltip
- Emacs keybindings in editor and console with checkbox toggle in Preferences->Keyboard Shortcuts
- Full support for enaml files, including introspection and an icon (based on the Cython icon)
- New SpyderInteractiveConsole, to be used as sys.excepthook, which starts code.interact with locals at the time of the exception, with hooks to the Spyder monitor.  The user can also put: `import code; code.interact()` and drop into the SpyderInteractiveConsole.
- Allow files to be dropped into Variable Explorer + support for .npz files
- Created SpyderActionPlugin for Plugins that do not provide docks + autopep8 example

**Comment by spyder-issues-migrator** [comment]
_From [mail.t...@gmx.de](https://code.google.com/u/104947229556192903674/) on 2013-11-10T09:51:55Z_

I am using this branch (oopps, clone), and it quite nice. Goodwork!

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-11-11T07:10:17Z_

Thanks!  Have you tried the "extras" bookmark?

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2013-11-13T06:20:15Z_

Great work Steven! There is a lot of work on both of your bookmarks.

I think your Jedi bookmark will have to wait until 2.4, because I'm still a bit worried about the loading times you report for numpy and friends, which were fixed in 2.2 after a big effort on my side. I also believe Jedi is the future but performance is paramount for us (especially in such user-facing things as code completion).

However, I think most of your Extras bookmark is ready for 2.3! I'm intrigued by SpyderInteractiveConsole, which sounds like Matlab's keyboard feature. If that's the case, then it'll be really really cool. Emacs keybindings and tab completions seem awesome too.

Let me release 2.3beta2, after which I plan to move to Bitbucket. Then I'd like to ask you a favor: please break this bookmark in several pull requests which we can review and discuss one by one in depth there.

Again, thanks for all your hard work and sorry for my huge delay in coming back to you. But this is high quality work and highly appreciated.

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-11-13T16:48:00Z_

Thanks for the feedback Carlos.  

Yes, SpyderInteractiveConsole is inspired by the keyboard in Matlab.  It allows you to interact with locals without being encumbered by pdb's aliases and behavior, like the inability to refer to the last output as "_".

Sure, I can break it up for Bitbucket.

About the speed issues, it is only slow for the FIRST request on numpy/qt/other large lib.  What I can do is make all of the introspection asynchronous, and load numpy at startup, and load selected libraries if they are imported in the current file.  Introspection requests would be buffered so there is only ever one in the queue, and would never block the UI.  I would suggest turning off auto-complete and auto-calltip since we would have Tab key completions.  Sound reasonable?

**Comment by spyder-issues-migrator** [comment]
_From [david.ve...@gmail.com](https://code.google.com/u/108342766018366950755/) on 2013-11-19T12:06:50Z_

I've been playing around with the extras bookmark and although I haven't done any extensive tests or comparisons yet it looks very good and stable!

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-11-19T18:02:48Z_

Thanks David, how has the "slowness" affected you, if at all?

**Comment by spyder-issues-migrator** [comment]
_From [david.ve...@gmail.com](https://code.google.com/u/108342766018366950755/) on 2013-11-20T01:11:15Z_

I can't notice any slowness at all...but my current machine has a blazing fast SSD with 8 Intel cores and 8GB or ram, so that might be part of the reason. Spyder dev with rope and the jedi branch seem to have the same speed for me. I'll try to get some more testing done on a slower machine I have at home (no SSD and older dual core AMD CPU, but also an Arch Linux box).

**Comment by spyder-issues-migrator** [comment]
_From [david.ve...@gmail.com](https://code.google.com/u/108342766018366950755/) on 2013-11-21T12:10:10Z_

Ok, some more testing on my 5 year old budget machine (AMD Atlhon dual core 4850, 2GB PC6400 RAM, standard HDD, up to date Arch Linux box with Jedi 0.7):
* first code completion invocation is slow on pylab, but not smaller modules
* later code completions are fast and almost instantaneous
* jedi feels slightly faster compared to rope
* when jedi is not installed, fallback to rope works as expected
* for a compiled library, such as opencv, jedi works perfectly, no code completion available with rope
* Some quick testing with drag-and-drop of files in the variable explorer did not work and a pop-up window with the following message appeared: 
Unable to load 'file:///home/dave/openmdao_log.txt'
Error message:
[Errno 2] No such file or directory: u'file:///home/dave/openmdao_log.txt'
If that is not a helpful message let me know than I'll try to do some debugging on a later date.
* I haven't looked at SpyderInteractiveConsole and SpyderActionPlugin (the autopep8 is an interesting feature), but I am not sure I understand how I could currently use them from within Spyder.

It looks like this branch is performing rock solid for me on Linux. As far as I am concerned, the initial slow cold start of Jedi is a minor issue compared to the extra features (I am thinking about code completion on compiled libraries now). 

One correction I have to make with respect to my previous post: the cold start effect is also present on my new 8 core machine with SSD, although the time delay on the first invocation is smaller compared my older desktop, it is still noticeable but not annoying.

Overall I would like to say: very cool and thumbs up :-)

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2013-11-21T12:27:03Z_

Really nice! If there is a fallback to rope (which I missed from Steven comments) then I think we could merge Jedi support in 2.3!

Besides, as Steven mentioned, we could avoid the cold start problem by making Jedi preload the main scientific modules in threads/QThreads while Spyder is starting.

The move to Bitbucket is almost ready (I hope to finish it tomorrow or during the weekend), so we could start to merge all this great work next week.

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-11-21T17:12:48Z_

Wow, thanks again David.  I removed the plugins temporarily, as they were interfering with my install and need more fleshing out.  I will take a look at the drag and drop.  

Woohoo!, thanks Carlos, I look forward to it.

**Comment by spyder-issues-migrator** [comment]
_From [contrebasse](https://code.google.com/u/contrebasse/) on 2013-11-21T23:54:33Z_

A word on autopep8 (I'm the original author for this plugin) : it adds an entry in the Source menu that autoformats your code (selection only if there's any, otherwise the whole current file).
You can install it on current spyder 2.3 (it requires py3compat) : https://github.com/Nodd/spyder_autopep8

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-11-23T07:23:08Z_

Sorry Joseph, I did not mean to cast disparity on your plugins.  I meant that my implementation with the SpyderActionPlugin was not working.  I had removed the LineProfile plugin because it was not part of my contribution in this repo.

David, I updated the Drop behavior, if you'd like to have a look.  I tested it on Windows and Ubuntu this time.

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-11-23T12:24:25Z_

Success!  Thanks to Joseph's fine example, I was able to get SpyderActionPlugin working with autopep8.  One only has to define a CONF_SECTION, CONFIGWIDGET_CLASS, and a register_plugin method, and you can be off and running.

**Comment by spyder-issues-migrator** [comment]
_From [contrebasse](https://code.google.com/u/contrebasse/) on 2013-11-23T12:31:05Z_

No problem Steven, I was answering to David's post where he said that he couldn't find how to use autopep8 in spyder. Your SpyderActionPlugin will be perfect for autopep8 !

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-11-24T13:34:27Z_

More improvements!  I stopped short of a fully asynchronous introspection, because I did not want to manage changed state between when a request was made and the result was available.  

The current version (in the extras bookmark) will preload numpy and matplotlib.pyplot, and then dynamically load the top level extension modules defined in  utils/modulecompletion if they are imported in the current file.  The current state of loading modules is shown in the status bar.  I think showing the module loading is a great way to manage the expectations of the user so they are not left wondering what the hell is going on when they do not get a completion. 

I still think auto-introspection should be disabled by default, as it can be disruptive to workflow in large modules (when using either jedi or rope).  

Any feedback is welcome!

**Comment by spyder-issues-migrator** [comment]
_From [david.ve...@gmail.com](https://code.google.com/u/108342766018366950755/) on 2013-11-26T00:57:14Z_

Did some very quick testing results on the drag-and-drop. This looks great and can become a killer feature in my opinion :-) Also the import helper when dropping files with a .txt extension is a very good idea. Some quick remarks, I hope to spent some more time reviewing this at another point in the future...:
* might be difficult to get that right in general, but dropping text files without *.txt are just ignored
* .pkl (Python pickled files) are ignored
* .mat files worked, but I think the way they are imported is a bit too nested and verbose. It might be better to resolve this upstream with SciPy (maybe they have actually good reasons for doing so), or maybe not.

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-11-26T03:19:35Z_

David, I agree on all counts.  What you are seeing is the default behavior that was already available by clicking the "Import Data" button in the variable explorer.  I had added .npz, and now .pkl based on your suggestion, but I did not want to change any other default behavior.  
Carlos, what are your thoughts?  Should we use the text import wizard for all files (after all, the user should know what they are doing)?  Also, can I tweak the way .mat files are brought in from MAT files similar to what I did in oct2py? http://pythonhosted.org/oct2py/conversions.html

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-11-26T03:23:41Z_

I meant to say use the text import wizard as the importer for new option, "* files" (do not exclude files without a file type.

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2013-11-27T13:43:57Z_

Steven, I already made official our move to Bitbucket. Our new address is: https://bitbucket.org/spyder-ide/spyderlib Please start to move your contributions there and make pull requests against that repo.

Let's start with your Jedi work so we can integrate it for beta3 (I'm about to release beta2 once I finish a couple of details).

From your extras bookmark, I would like to merge first these additions:

1. Tab completions and re-trigger calltip
2. Full support for enaml files
3. Enter post-mortem mode on tracebacks.
4. Variable-explorer drag and drop.

Please open a separate pull request for each one.


About comment 58: Yeah, please improve our MAT import facilities as you wish, this is an area we haven't touched in a long time.

About comment 59: You mean to offer the "Import from clipboard" dialog? and paste there the file a user wants to import?

**Labels:** -MS-v2.4 -Cat-Editor MS-v2.3 Cat-CodeCompletion

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-11-27T14:50:13Z_

Carlos,

Oh boy, here I come.

I will wait on the MAT file improvements until after this round of PRs.

I am going to add a json import handler right now since that is a quick one.

About comment 59, I mean that we should be able to open the text import wizard (the one that is brought up when a text file is imported in the Variable Explorer) for *all* files that are not explicitly handled by other import wizards, subject to a test for text-likeness of the content.
In the same token, it would be nice to drag any type of file into the editor that looks like a text file. http://stackoverflow.com/a/3002505

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-11-29T04:27:00Z_

David,

I added the enhanced Matlab import behavior to my bitbucket repo, in the "variable-explorer-drag-n-drop" branch, if you'd like to experiment with it. https://bitbucket.org/blink1073/spyderlib

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-11-29T04:56:04Z_

David,

I tested it it using the oct2py datatypes test file (link below) with the Octave commands:

test = test_datatypes
save -v6 "~/test.mat" test

You'll see the output for that file is a MatlabStruct, which is an object that is a dictionary, but supports attribute access, and auto-creation of sub-structures (like the Matlab struct). https://github.com/blink1073/oct2py/blob/master/oct2py/tests/test_datatypes.m

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-11-29T09:24:17Z_

More progress!  An editor will now accept an unknown file type drop, provided it passes the new utils.sourcecode.is_binary() test.  

The Variable Explorer accepts ANY file, and will ask you what to import it as if it does not support the file extension.

These changes are in the "variable-explorer-drag-n-drop" branch at: https://bitbucket.org/blink1073/spyderlib

**Comment by spyder-issues-migrator** [comment]
_From [david.ve...@gmail.com](https://code.google.com/u/108342766018366950755/) on 2013-12-01T14:00:21Z_

Sorry for the late reply. It seems to work quite nicely. Variables imported from the .mat workspace as one might expect: they are dropped right into the current workspace. I am actually using oct2py for the moment to launch a Matlab code in Octave via Python so I've noticed it worked quite well there, but I didn't realize you were the one to blame Steven, thanks :-) !! One small remark: I noticed the code completion in the console (both Python and IPython) wasn't able to give me any suggestions on the nested variables under test (so test.struct_array, test.logical, etc) even though they where defined and valid.

Quite nice to see how it handles unknown file-drag-and-drop too! Two small comments/suggestions:
* in the preview section: it might be me, but I wasn't sure what variable type it would use if import as array wasn't selected. When I saw it was a list it made sense, but it could be even more clear if you had to chose between either list or array. 
* Initially I missed the variable field name (maybe rename from "name" to "variable name"?). I realize this is just a matter of opinion, but maybe it is more visible if aligned to the left instead of the right? Also, longer file names are not visible unless you "scroll" through them with the arrows due to the short size (length wise) of the field.

Just because I was curious how it would handle big arrays, I created 4 different size=(100000, 1000) random float arrays in octave, saved them to a single mat file of roughly 3 GB (using "save -v6"). Drag-and-dropped in Spyder worked like a charm. I can safely say there is definitely no double/redundant allocation going on in memory on those arrays on Spyder's side :-)

Thanks for the work! I hope these comments are useful.

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-12-01T14:42:49Z_

Thanks again David.  I will have to think about the auto-complete, since a MatlabStruct is technically a dictionary, and I don't want to store the data in two places (in the dict and as an attribute).  I could invert the behavior and use it as a simple object with a dict-like interface.  I'll try it out and let you know.
 
The text import wizard was pre-existing, but I will make your suggested changes and let others veto them if they do not like them.

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-12-01T17:42:41Z_

*Awesome* David, that was a great idea.  You can now code complete (you have to contend with the dict built in names), and I added the ability to bring up the explorer window if you double click on it.  I will add this back in Oct2Py.  Keep up the great suggestions.

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-12-02T06:06:33Z_

David,

I believe I got all the upgrades, except I do not know what you mean by:
"Also, longer file names are not visible unless you "scroll" through them with the arrows due to the short size (length wise) of the field."
On which pane?

**Comment by spyder-issues-migrator** [comment]
_From [david.ve...@gmail.com](https://code.google.com/u/108342766018366950755/) on 2013-12-02T08:11:26Z_

\> I added the ability to bring up the explorer window if you double click on it
Nice and it just works!

I discovered an issue with the nested variable names though: I can see they exists for a minute or so, but than they disappear. It is as if the garbage collector removes the references to the nested ones? I can keep changing/calling the variables via the dictionary. Is that something you can also see on your end? I tried to allocate a different values to one of the nested variables, not sure if that triggered that behaviour for me.

In the preview pane you have the "variable name" (the one for which you now changed the alignment to the left) followed by a text field where you change the variable name once imported in the workspace. That field was a bit short by default, so I couldn't see the complete variable name if the imported text file name was long. Seems you made it wider now, correct? Maybe it could be as wide as the window itself and resize if the parent window is resized? I have no idea if this is easy to implement, if not it doesn't matter.

It is difficult to keep up just checking out the stuff you implement :-)

Cheers

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-12-02T17:43:23Z_

David,

I tried changing a nested value through the console and through the editor, and leaving it open for a few minutes, but the value did not disappear or revert.  I also added a key/value pair in the editor.  I'm not sure what to make of what you're seeing.  You have have hit Cancel instead of OK, and then reopened it.  I did that by accident once.

The variable name input now expands to fill the rest of the space, and with the window.

**Comment by spyder-issues-migrator** [comment]
_From [trueflyi...@gmail.com](https://code.google.com/u/103824930862744342842/) on 2013-12-03T05:57:47Z_

umm, sorry, but how does all that relate to Jedi autocompletion? could we move the discussion about matlab-compatibility and variable explorer functionality into other bug reports please?

no doubt that this is awesome, though, a huge argument to convert matlab users over and ease interoperability!

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-12-03T07:19:41Z_

Hey, I am having fun got carried away. I find Google Code Issues to be obnoxious.

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2013-12-03T07:32:05Z_

The I think it's better for you to open a PR and continue the discussion with David in Bitbucket (because at the end he's helping you to improve your code).

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-12-03T07:44:12Z_

Done.

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2013-12-03T08:05:35Z_

https://bitbucket.org/spyder-ide/spyderlib/pull-request/4/improved-drag-and-drop-behavior-in-editor/diff

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2013-12-16T08:08:40Z_

This issue was updated by revision 4bde999057ca .


- The idea of this PR was to add support for the Jedi completion library.
Unfortunately (right now) it consumes too much memory to preload
compiled libraries.
- So an important refactoring was done to support it once that problem
is solved.
- As part of this refactoring were also added token-based completions,
which are shown only when no rope-based ones can be found. They work
for any of file type too (html, css, etc).
- Finally, regexp-based goto-definitions were also added.

**Status:** Started

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2014-01-07T17:41:04Z_

Update: The author of Jedi is now actively working on a resolution to the memory problem.

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2014-04-16T08:57:11Z_

**Blocking:** spyderlib:1469

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2014-05-04T14:57:01Z_

Hopefully for 2.3.1, Jedi 0.8 has been released! (really waiting for it :)

**Labels:** -MS-v2.3 MS-v2.3.1

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2014-05-07T10:26:38Z_

It didn't take them too much. Two days ago Jedi 0.8 was released!

Steven, could you a prepare a PR to start testing it ASAP? Of course it won't be included in 2.3 (because we are about to release) but I hope there won't be major problems for 2.3.1.

**Owner:** steven.s...@gmail.com

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2014-05-07T18:16:48Z_

I'll take a stab it this weekend.

**Comment by spyder-issues-migrator** [comment]
_From [phyo.arkarlwin](https://code.google.com/u/phyo.arkarlwin/) on 2014-06-24T16:48:16Z_

So theres no Jedi support in current version right? I am going to test it. Gotta clone from steven's repo i guess.

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2014-06-24T19:19:55Z_

Right, clone my repo on bitbucket (blink1073/spyderlib) and checkout the jedi-0.8 branch.

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2014-07-10T19:24:37Z_

**Labels:** -MS-v2.3.1 MS-v2.4

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2014-11-18T15:28:51Z_

**Blocking:** spyderlib:1993

**Comment by spyder-issues-migrator** [comment]
_From [thomson....@gmail.com](https://code.google.com/u/110689693938517381157/) on 2014-12-03T19:38:40Z_

I came over here from this thread: https://groups.google.com/forum/#!topic/spyderlib/RFGB8JY0X3g I am curious about where we are with this. I use Anaconda, do a lot of PySide, and am not getting any code completion yet...

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2014-12-04T13:06:07Z_

I plan to merge this soon but it will only be available for 2.4.

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2014-12-20T17:19:22Z_

This issue was closed by revision 132d65d01f96 .

**Status:** Fixed

**Comment by spyder-issues-migrator** [comment]
_From [steven.s...@gmail.com](https://code.google.com/u/118403364018775626054/) on 2014-12-20T18:09:34Z_

YES!!!!!!  Only took me two years  :)

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2014-12-21T12:10:27Z_

Damn! True ;-) But we also needed Jedi to mature much more than what it was two years ago.

**Comment by spyder-issues-migrator** [comment]
_From [david.ve...@gmail.com](https://code.google.com/u/108342766018366950755/) on 2014-12-22T07:27:03Z_

Job well done!! :-) Patience still is a virtue ;-)

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2015-02-08T09:08:25Z_

More good news: I decided to backport the fantastic work of Steven to 2.3.3, so that Ubuntu users in Python 3 can have code completion for 15.04 ;-) (given that there is no python3-rope package, and I just noticed it a couple of months ago!)

**Labels:** -MS-v2.4 MS-v2.3.3
